File: publisher.cc
// https://ignitionrobotics.org/api/transport/11.0/messages.html

#include <atomic>
#include <chrono>
#include <csignal>
#include <iostream>
#include <string>
#include <thread>
#include <ignition/msgs.hh>
#include <ignition/transport.hh>

static std::atomic<bool> g_terminatePub(false);

void signal_handler(int _signal)
{
  if (_signal == SIGINT || _signal == SIGTERM)
    g_terminatePub = true;
}

int main(int argc, char **argv)
{
  // Install a signal handler for SIGINT and SIGTERM.
  std::signal(SIGINT,  signal_handler);
  std::signal(SIGTERM, signal_handler);
  
  // Create a transport node and advertise a topic.
  ignition::transport::Node node;
  std::string topic = "/foo";
  auto pub = node.Advertise<ignition::msgs::StringMsg>(topic);
  if (!pub)
  {
    std::cerr << "Error advertising topic [" << topic << "]" << std::endl;
    return -1;
  }
  
  // Prepare the message.
  ignition::msgs::StringMsg msg;
  msg.set_data("hello");

  // Publish messages at 1Hz.
  while (!g_terminatePub)
  {
    if (!pub.Publish(msg))
    {
      break;
    }
    std::cout << "Publishing hello on topic [" << topic << "]" << std::endl;
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
  }
  return 0;
}

File: python/publisher.py
#!/usr/bin/env python

# Copyright (C) 2022 Rhys Mainwaring
#
# Licensed under the Apache License, Version 2.0 (the "License")
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#       http:#www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import time

from gz.msgs.stringmsg_pb2 import StringMsg

from gz.transport import AdvertiseMessageOptions
from gz.transport import Node

def main():
    # Create a transport node and advertise a topic
    node = Node()
    topic = "/foo"
    msg_type_name = StringMsg.DESCRIPTOR.full_name
    pub_options = AdvertiseMessageOptions()
    pub = node.advertise(topic, msg_type_name, pub_options)
    if pub.valid():
        print("Advertising {} on topic [{}]".format(msg_type_name, topic))
    else:
        print("Error advertising topic [{}]".format(topic))

    # Prepare the message
    msg = StringMsg()
    msg.data = "hello"

    try:
        while True:
          if not pub.publish(msg):
              break

          print("Publishing hello on topic [{}]".format(topic))
          time.sleep(1.0)

    except KeyboardInterrupt:
        pass

if __name__ == "__main__":
    main()

